chia/full_node/full_node.py

```diff
@@ -531,6 +532,8 @@ def _num_needed_peers(self) -> int:
 
     def _close(self):
         self._shut_down = True
+        if self._init_weight_proof is not None:
+            self._init_weight_proof.cancel()
```

Comment:
not sure cancelling the task will break the loop inside create_sub_epoch_segments 
i think we need to add self.close:bool  to the WeightProofHandler so we can break if we are on shutdown and loop is still running

Reply:
it will make the next `await` throw a `CancellationException`, as far as I understand. If there are places that can't fail with exceptions without corrupting the database it would probably be best to handle those. I'll around

Reply:
I can't find any place where it would be obviously incorrect to cancel at an `await` site. Also, I don't think this would make it any worse than what it already is, where the task is interrupted like this instead:

```
future: <Task finished name='Task-4' coro=<FullNode.initialize_weight_proof() done, defined at /home/arvid/dev/chia-blockchain/chia/full_node/full_node.py:170> exception=ProgrammingError('Cannot operate on a closed database.')>
Traceback (most recent call last):
  File "/home/arvid/dev/chia-blockchain/chia/full_node/full_node.py", line 174, in initialize_weight_proof
    await self.weight_proof_handler.create_sub_epoch_segments()
  File "/home/arvid/dev/chia-blockchain/chia/full_node/weight_proof.py", line 253, in create_sub_epoch_segments
    await self.__create_persist_segment(prev_ses_block, ses_block, ses_height, sub_epoch_n)
  File "/home/arvid/dev/chia-blockchain/chia/full_node/weight_proof.py", line 260, in __create_persist_segment
    segments = await self.blockchain.get_sub_epoch_challenge_segments(ses_block.header_hash)
  File "/home/arvid/dev/chia-blockchain/chia/consensus/blockchain.py", line 726, in get_sub_epoch_challenge_segments
    segments: Optional[List[SubEpochChallengeSegment]] = await self.block_store.get_sub_epoch_challenge_segments(
  File "/home/arvid/dev/chia-blockchain/chia/full_node/block_store.py", line 122, in get_sub_epoch_challenge_segments
    await cursor.close()
  File "/home/arvid/dev/chia-blockchain/venv/lib/python3.8/site-packages/aiosqlite/cursor.py", line 69, in close
    await self._execute(self._cursor.close)
  File "/home/arvid/dev/chia-blockchain/venv/lib/python3.8/site-packages/aiosqlite/cursor.py", line 31, in _execute
    return await self._conn._execute(fn, *args, **kwargs)
  File "/home/arvid/dev/chia-blockchain/venv/lib/python3.8/site-packages/aiosqlite/core.py", line 129, in _execute
    return await future
  File "/home/arvid/dev/chia-blockchain/venv/lib/python3.8/site-packages/aiosqlite/core.py", line 102, in run
    result = function()
```

Reply:
that actually sounds like it would work fine, Thanks